ccp/src/hashrate/collector.rs

```diff
@@ -0,0 +1,220 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::time::Instant;
+
+use ccp_shared::types::EpochParameters;
+use ccp_shared::types::LogicalCoreId;
+
+use super::record::HashrateRecordType;
+use super::record::ThreadHashrateRecord;
+
+/// Collects and analyzes hashrate comes from sync threads.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct HashrateCollector {
+    status: CollectorStatus,
+    entries: HashMap<LogicalCoreId, ThreadHashrateRaw>,
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub(crate) enum CollectorStatus {
+    Busy {
+        started_time: Instant,
+        epoch: EpochParameters,
+    },
+    #[default]
+    Idle,
+}
+
+pub(crate) type Hashrate = HashMap<LogicalCoreId, ThreadHashrate>;
+
+/// Unprocessed cumulative hashrate for a sync thread.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub(crate) struct ThreadHashrateRaw {
+    cache_creation: ParameterStatus<Duration>,
+    dataset_initialization: ParameterStatus<Duration>,
+    cc_job_duration: ParameterStatus<Duration>,
+    checked_hashes_count: u64,
+    found_proofs_count: u64,
+}
+
+/// Processed cumulative hashrate for a sync thread.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct ThreadHashrate {
+    // hashes from epoch start, counts operations with cache and dataset
+    pub(crate) effective_hashrate: f64,
+    // pure hashrate, which doesn't count operations with cache and dataset
+    pub(crate) hashrate: ParameterStatus<f64>,
+    pub(crate) proofs_found: u64,
+    pub(crate) cache_creation: ParameterStatus<Duration>,
+    pub(crate) dataset_initialization: ParameterStatus<Duration>,
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) enum ParameterStatus<T> {
+    Measured(T),
+    #[default]
+    NotMeasured,
+}
+
+#[derive(Clone, Debug, Default)]
+pub(crate) enum EpochObservation {
+    EpochChanged {
+        prev_epoch_hashrate: Hashrate,
+    },
+    #[default]
+    EpochNotChanged,
+    StartedWorking,
+}
+
+impl HashrateCollector {
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn account_record(
+        &mut self,
+        hashrate_record: ThreadHashrateRecord,
+    ) -> EpochObservation {
+        use std::collections::hash_map::Entry;
+
+        let result = self.observe_epoch(hashrate_record.epoch);
+
+        match self.entries.entry(hashrate_record.core_id) {
```

Comment:
`entry(...).or_default().accout_record(hashrate_record)`.

Reply:
I wish clippy be able to make warnings about that.